File: src/CI3.java
import java.util.ArrayList;
import java.util.List;

public class CI3 {

    public static void main(String args[]) {
        System.out.println("-- start --");
        List<Integer> result = getDuplicateNumber(new int[]{2, 3, 1, 0, 2, 5, 3});
        if (result != null) {
            result.forEach(System.out::println);
        }
        System.out.println("-- end --");

        int oneDuplicateNumber = getOneDuplicateNumber(new int[]{2, 3, 5, 4, 3, 2, 6, 7});
        System.out.println("{2, 3, 5, 4, 3, 2, 6, 7} oneDuplicateNumber = " + oneDuplicateNumber);


        int oneDuplicateNumber2 = getOneDuplicateNumber(new int[]{2, 2, 5, 4, 3, 2, 6, 7});
        System.out.println("{2, 2, 5, 4, 3, 2, 6, 7} oneDuplicateNumber = " + oneDuplicateNumber2);


        int oneDuplicateNumber3 = getOneDuplicateNumber(new int[]{1, 2, 3, 2});
        System.out.println("{1, 2, 3, 2} oneDuplicateNumber = " + oneDuplicateNumber3);


        int oneDuplicateNumber4 = getOneDuplicateNumber(new int[]{1, 2});
        System.out.println("{1,2} oneDuplicateNumber = " + oneDuplicateNumber4);

        int oneDuplicateNumber5 = getOneDuplicateNumber(new int[]{1, 2, 3, 4});
        System.out.println("{1,2，3，4} oneDuplicateNumber = " + oneDuplicateNumber5);

        int oneDuplicateNumber6 = getOneDuplicateNumber(new int[]{2, 3, 5, 4, 3, 2, 6});
        System.out.println("{2, 3, 5, 4, 3, 2, 6} oneDuplicateNumber = " + oneDuplicateNumber6);

    }


    private static List<Integer> getDuplicateNumber(int[] inputArray) {
        if (inputArray == null || inputArray.length == 0) {
            return null;
        }
        int length = inputArray.length;
        List<Integer> result = new ArrayList<>();

        for (int i = 0; i < length; i++) {
            while (inputArray[i] != i) {
                if (inputArray[inputArray[i]] == inputArray[i]) {
                    result.add(inputArray[i]);
                    break;
                } else {
                    int temp = inputArray[i];
                    inputArray[i] = inputArray[temp];
                    inputArray[temp] = temp;
                }
            }
        }
        return result;
    }


    private static int getOneDuplicateNumber(int[] inputArray) {
        if (inputArray == null || inputArray.length == 0 || inputArray.length == 1) {
            return -1;
        }

        int length = inputArray.length;
        int start = 1;
        int end = length - 1;

        while (end >= start) {
            int middle = (start + end) / 2;
            int countInRange = countRange(inputArray, length, start, middle);
            //分两种情况 有可能最后两个指针重合，也有可能最后两个指针相差1
            if (end == start) {
                if (countInRange > 1) {
                    return start;
                } else {
                    break;
                }
            } else if (end == start + 1) {
                if (countInRange > 1) {
                    return start;
                } else if (countRange(inputArray, length, end, end) > 1) {
                    return end;
                } else {
                    break;
                }
            }

            //二分查找，折半缩小范围
            if (countInRange <= middle) {
                start = middle + 1;
            } else {
                end = middle;
            }
        }
        return -1;
    }

    private static int countRange(int[] inputArray, int length, int start, int end) {
        if (inputArray == null) {
            return 0;
        }
        int count = 0;
        for (int i = 0; i < length; i++) {
            if (inputArray[i] >= start && inputArray[i] <= end) {
                count++;
            }
        }
        return count;
    }

}

File: src/CI3.kt

fun main(args: Array<String>) {
    println("-- begin --")
    val result = getDuplicateNumber(intArrayOf(2, 3, 1, 0, 2, 5, 3))
    result?.forEach { println(it) }
    println("-- end --")

    val oneDuplicateNumber = getOneDuplicateNumber(intArrayOf(2, 3, 5, 4, 3, 2, 6, 7))
    println("{2, 3, 5, 4, 3, 2, 6, 7} oneDuplicateNumber = $oneDuplicateNumber")

    val oneDuplicateNumber2 = getOneDuplicateNumber(intArrayOf(2, 2, 5, 4, 3, 2, 6, 7))
    println("{2, 2, 5, 4, 3, 2, 6, 7} oneDuplicateNumber = $oneDuplicateNumber2")


    val oneDuplicateNumber3 = getOneDuplicateNumber(intArrayOf(1, 2, 3, 2))
    println("{1, 2, 3, 2} oneDuplicateNumber = $oneDuplicateNumber3")


    val oneDuplicateNumber4 = getOneDuplicateNumber(intArrayOf(1, 2))
    println("{1,2} oneDuplicateNumber = $oneDuplicateNumber4")

    val oneDuplicateNumber5 = getOneDuplicateNumber(intArrayOf(1, 2, 3, 4))
    println("{1,2，3，4} oneDuplicateNumber = $oneDuplicateNumber5")

    val oneDuplicateNumber6 = getOneDuplicateNumber(intArrayOf(2, 3, 5, 4, 3, 2, 6))
    println("{2, 3, 5, 4, 3, 2, 6} oneDuplicateNumber = $oneDuplicateNumber6")

}

private fun getDuplicateNumber(inputArray: IntArray?): List<Int>? {
    if (inputArray?.isEmpty()!!) {
        return null
    }
    val length = inputArray.size
    val result = ArrayList<Int>()

    for (i in 0 until length) {
        while (inputArray[i] != i) {
            if (inputArray[inputArray[i]] == inputArray[i]) {
                result.add(inputArray[i])
                break
            } else {
                val temp = inputArray[i]
                inputArray[i] = inputArray[temp]
                inputArray[temp] = temp
            }
        }
    }
    return result
}


fun getOneDuplicateNumber(inputArray: IntArray?): Int {
    if (inputArray?.isEmpty()!!) {
        return 0
    }
    val length = inputArray.size
    var start = 1
    var end = length - 1

    while (start <= end) {
        val middle = (start + end) / 2
        val count = countRange(inputArray, start, middle)
        if (start == end) {
            if (count > 1) {
                return start
            } else {
                break
            }
        } else if (start + 1 == end) {
            return if (count > 1) {
                start
            } else if (countRange(inputArray, end, end) > 1) {
                end
            } else {
                break
            }
        }

        if (count <= middle) {
            start = middle + 1
        } else {
            end = middle
        }
    }
    return -1
}


fun countRange(inputArray: IntArray?, start: Int, end: Int): Int {
    if (inputArray?.isEmpty()!!) {
        return 0
    }
    var count = 0
    inputArray.forEach {
        if (it in start..end) {
            count++
        }
    }
    return count
}
